pilot/pkg/proxy/envoy/xds/xds.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"net"
+
+	"google.golang.org/grpc"
+	"istio.io/istio/pilot/pkg/config/memory"
+	"istio.io/istio/pilot/pkg/model"
+	v2 "istio.io/istio/pilot/pkg/proxy/envoy/v2"
+	"istio.io/istio/pilot/pkg/serviceregistry"
+	"istio.io/istio/pilot/pkg/serviceregistry/aggregate"
+	"istio.io/istio/pilot/pkg/serviceregistry/external"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/mesh"
+	"istio.io/istio/pkg/config/schema/collections"
+)
+
+type Server struct {
+	// DiscoveryServer is the gRPC XDS implementation
+	// Env and MemRegistry are available as fields, as well as the default PushContext.
+	DiscoveryServer *v2.DiscoveryServer
+
+	// MemoryStore is an in-memory config store, part of the aggregate store used by the discovery server.
+	MemoryConfigStore model.IstioConfigStore
+
+	AggregateConfigControoler *aggregate.Controller
+	GRPCListener              net.Listener
+}
+
+// Creates an basic, functional discovery server, using the same code as Istiod, but
+// backed by an in-memory config and endpoint stores.
+//
+// Can be used in tests, or as a minimal XDS discovery server with no dependency on K8S and
+// the complex bootstrap used by Istiod.
+func NewXDS() *Server {
```

Comment:
Should this be in a test package?

Reply:
See the second part of the comment - I'm starting it in istio-agent ( which already runs SDS server ), as part of the prototype, and to avoid having 2 connections to XDS.
So Agent will connect to XDS - grab API listeners for DNS - and keep the connection open, with Envoy connecting to localhost. 

Highly experimental - but we want to get feedback.